Use MenuCategoryData to generate array in TopMenu plugin to allow extension via plugins

fredden opened the pull request:
### Description
Following [discussion on Magento Community Engineering Slack instance](https://magentocommeng.slack.com/archives/CBSL1DF8B/p1592055254155800), it has become clear that the `private` method `Magento\Catalog\Plugin\Block\Topmenu::getCategoryAsArray()` should be abstracted to allow extension via plugin (as inheritance (via preference) is to be discouraged). It appears that an existing class provides most of the functionality required. This pull request changes `getCategoryAsArray()` to use this existing external class (with `public` method) to allow this data to be augmented via plugin on the latter.

### Related Pull Requests
*None known*

### Fixed Issues
*None known*

### Manual testing scenarios
See also discussion on Slack (link above).

1. Attempt to add an item to output of `getCategoryAsArray()` from external code (third party extension)

### Questions or comments
I have not yet adjusted any tests, however this should not result in any change in functionality.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29664: Use MenuCategoryData to generate array in TopMenu plugin to allow extension via plugins

fredden committed:
Use existing external class to generate array

This is in order to allow additions to the return value of the private method
getCategoryAsArray() via a plugin on
Magento\Catalog\Observer\MenuCategoryData::getMenuCategoryData()

lbajsarowicz reviewed (changes requested):
Injection of `Observer` to `Plugin` is not the best idea at all.
If you need the logic reusable across these two - just create separate class that provides Interface you need.

fredden committed:
Move model out of 'Observer' namespace

fredden commented:
`MenuCategoryData` isn't an observer, but a model in the wrong namespace. I've now moved it to 'Model'.

sidolov commented:
@magento create issue

fredden committed:
Merge branch '2.4-develop' into topmenu-extensibility

fredden commented:
@magento run all tests

fredden committed:
Merge branch '2.4-develop' into topmenu-extensibility

fredden commented:
@magento run all tests

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

fredden committed:
Import classes where needed

fredden committed:
Declare parameter and return types

fredden committed:
Fix tests

fredden commented:
@magento run all tests

lbajsarowicz commented:
@magento run all tests

fredden commented:
@magento run all tests

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

fredden commented:
@magento run all tests

fredden committed:
Merge branch '2.4-develop' into topmenu-extensibility

fredden commented:
@magento run all tests

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

fredden commented:
@magento run all tests

fredden committed:
Fix static code analysis tests

fredden commented:
@magento run all tests

fredden commented:
@magento run all tests

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

fredden commented:
@magento run all tests

fredden committed:
Merge branch '2.4-develop' into topmenu-extensibility

fredden commented:
@magento run all tests

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

fredden commented:
@magento run all tests

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

engcom-Hotel committed:
Merge branch '2.4-develop' into topmenu-extensibility

engcom-Hotel commented on app/code/Magento/Catalog/Model/MenuCategoryData.php:
```diff
@@ -3,45 +3,45 @@
  * All Rights Reserved.
  */
-namespace Magento\Catalog\Observer;
+namespace Magento\Catalog\Model;
```
Updating namespace might be a breaking change for the third party extensions or modules directly referencing `Magento\Catalog\Observer\MenuCategoryData` will break.

Leave the class empty in observer namespace by extending this class.

fredden replied:
This seems fair. I have added a class as suggested in 80a8a30282279ae1983c9a1787c8ddb1c9285f9e.

engcom-Hotel reviewed (changes requested):
Hello @fredden,

Thank you for your contribution!

Please fix the failed static and integration tests.

Also refer to the below review comment.

Thank you

fredden committed:
Add class to avoid breaking backward compatibility

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

fredden committed:
Reduce visibility to appease static linter tool

fredden committed:
Update copyright year to appease static linter

engcom-Hotel reviewed (changes requested):
Hello @fredden,

Thank you for making the changes!

The static and integration are still failing. Please fix those as well.

Thank you

fredden committed:
Update copyright string to appease linter

fredden commented:
@magento run all tests

fredden committed:
Enable strict_types to appease linter

fredden commented:
@magento run all tests

fredden committed:
Explicitly cast values to integers

These methods should return integers according to their type information,
however in practice we are sometimes getting string values come back. The
proper fix for this would be to move the type information from the docblock to
PHP native types, which would enforce the declared return type. Doing that is
out of scope for this task.

fredden commented:
@magento run all tests

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

engcom-Hotel reviewed (changes requested):
Hello @fredden,

Thank you for updating the PR!

We are still seeing some static test failure. Please resolve them.

engcom-Hotel commented on app/code/Magento/Catalog/Model/MenuCategoryData.php:
```diff
@@ -0,0 +1,97 @@
+
+        $isActiveCategory = false;
+        /** @var Category $currentCategory */
+        $currentCategory = $this->registry->registry('current_category');
```
> ## Magento Health Index
> <img alt="Screenshot_2025-12-18_10-13-31" width="955" height="238" src="https://private-user-images.githubusercontent.com/334786/528032418-ba2c16fc-5d58-40ff-b2e4-10d1530699b6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY0MDc0NTIsIm5iZiI6MTc2NjQwNzE1MiwicGF0aCI6Ii8zMzQ3ODYvNTI4MDMyNDE4LWJhMmMxNmZjLTVkNTgtNDBmZi1iMmU0LTEwZDE1MzA2OTliNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIyMlQxMjM5MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZDhmN2NkOTBlYWNmZDgxNDFmOGEzNDdkZWNiZGU0NTAzOTk0MDU0YWEwMzMyMzcyZjczZDRkZDZjYTY1Zjc5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bOixqUQgmqP7BY_oVbga9f7nBfqcDkbjKLozyPyT7Qg">
> https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/28715/90f7639e613950ea7dca8b428245ff33/MagentoHealthIndex/report.html
> 
> From what I can tell, this check is complaining that there is some use of [PHP superglobals](https://www.php.net/manual/en/language.variables.superglobals.php) in `app/code/Magento/Catalog/Model/MenuCategoryData.php`. I can't see any such usage. Please can you show me what line number(s) this check is complaining about?

The static analysis tool is flagging the use of `Magento\Framework\Registry` as global state usage. The Registry pattern is deprecated in Magento 2.

I guess as a solution `hasActive()` method correctly uses the Layer resolver to get the current category:
`$currentCategory = $catalogLayer->getCurrentCategory();`

fredden replied:
@engcom-Hotel please can you work to get this error resolved in the source file: https://github.com/magento/magento2/blob/2.4.9-alpha3/app/code/Magento/Catalog/Observer/MenuCategoryData.php#L63 (in a different pull request / internally).
I would prefer not to make such changes in this pull request as that is scope creep. Please can you ensure that the codebase adheres to the current coding standards. Situations like this generate significant friction for contributors.

engcom-Dash committed:
Merge branch '2.4-develop' into topmenu-extensibility

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into topmenu-extensibility

engcom-Hotel committed:
Merge branch '2.4-develop' into topmenu-extensibility

engcom-Hotel committed:
Merge branch '2.4-develop' into topmenu-extensibility

engcom-Charlie committed:
Fixing static test failures

engcom-Charlie committed:
Merge branch '2.4-develop' into topmenu-extensibility

fredden commented:
> I have fixed the static test failures. To move forward with this PR, we need the MHI to be fixed in this PR. Should I go ahead and take care of that?

No, the "MHI" (which I'm choosing to read as "Magento Health Index") check failure should NOT be fixed here. It is complaining about code that is NOT being changed in this pull request. See https://github.com/magento/magento2/pull/28715#discussion_r2649114472 where I asked for this to be fixed in a separate pull request by your team; you have added this check but have not done the necessary work to make your existing codebase comply with the rules being imposed. When the actual issue is resolved, the check will pass here.
Making changes to how the registry is used is out of scope for this pull request and will only confuse matters if included.

engcom-Charlie commented:
In order to proceed further on this PR the MHI needs to be fixed. Since there is no activity on [this](https://github.com/magento/magento2/pull/28715#issuecomment-4332932111) we are closing the PR now. Please feel free to reopen on any other further updates.
